In [ ]:
import psycopg2
import pandas as pd

from sqlalchemy import create_engine, text as sql_text

In [ ]:
%load_ext sql

In [ ]:
%sql postgresql://tttqkwhz:WLOVMdNsIUWNmq6J0sIRuAInQKg0QDs7@mel.db.elephantsql.com:5432/tttqkwhz

## Question 1

Untuk mengetahui target customer BeeCycle.
Kamu melakukan query untuk melihat distribusi customer berdasarkan gender dan umur.

Dimana untuk category umur, kamu membagi umur customer kedalam  (Hint : CASE WHEN)
* umur customer <= 20 tahun maka **'Group <=20'**
* umur customer antara 21 dan 40 tahun maka **'Group 21 - 40'**
* umur customer antara 41 dan 60 tahun maka **'Group 41 - 60'**
* umur customer lebih dari 60 tahun maka **'Group >60'**

Dengan grouping_age yg sudah didefine, selanjutnya jawab pertanyaan berikut

**Grouping age apa dan gender apa yg memiliki transaksi paling tinggi di BeeCycle?**


In [ ]:
#Tulis Query disini
%%sql
with age_group as(
	select customer_id, gender,
	case
	when age <= 20 then 'Group<=20'
	when age > 20 and age <= 40 then 'Group 21-40'
	when age > 40 and age <= 60 then 'Group 41-60'
	else 'Group >60'
	end as grouping_age
	from dim_customer dc
	)

select
ag.grouping_age,
ag.gender,
count(fc.customer_id) as total_transaction
from fact_sales fc
join age_group ag on ag.customer_id=fc.customer_id
group by 1,2
order by 3 desc

 * postgresql://tttqkwhz:***@mel.db.elephantsql.com:5432/tttqkwhz
6 rows affected.


[('Group 21-40', 'F', 1551),
 ('Group 41-60', 'F', 1454),
 ('Group 21-40', 'M', 1401),
 ('Group 41-60', 'M', 1304),
 ('Group >60', 'M', 161),
 ('Group >60', 'F', 82)]

**Grouping age apa dan gender apa yg memiliki transaksi paling tinggi di BeeCycle?**\
Jawaban: "Group 21-40	F"	dengan total transaksi sebanyak 1551

## Question 2


Karena bulan Oktober depan akan ada campaign.
Kamu punya ide untuk menghilight warna product tertentu yg menjadi kegemaran customer. Kamu selanjutnya melakukan query untuk menjawab pertanyaan berikut :



**Warna apa di tiap tahun yg menjadi warna paling populer dibeli oleh customer?**

In [ ]:
#Tulis Query disini

%%sql
with color_ranking as (
  select
    color,
    date_part('year', date(fc.order_date)) as date_year,
    count(dp.product_id) as total_colour_sold,
    dense_rank() over (partition by date_part('year', date(fc.order_date))
                      order by count(dp.product_id) desc) as ranking
  from fact_sales fc
  join dim_product dp on dp.product_id = fc.product_id
  where color != 'NA'
  group by 1, 2
)

select * from color_ranking where ranking = 1

 * postgresql://tttqkwhz:***@mel.db.elephantsql.com:5432/tttqkwhz
4 rows affected.


[('Red', 2016.0, 216, 1),
 ('Red', 2017.0, 223, 1),
 ('Black', 2018.0, 503, 1),
 ('Black', 2019.0, 281, 1)]

**Warna apa di tiap tahun yg menjadi warna paling populer dibeli oleh customer?**\
Jawaban: Tahun 2016 dan 2017 adalah warna merah (red), tahun 2018 dan 2019 warna hitam (red)

## Question 3


Karena campaign bulan Oktober tsb hanya terbatas untuk tiap2 toko cabang.
kamu punya ide, bagaimana hanya mengambil TOP 10 product_id dari masing2 toko cabang. sehingga kamu melakukan query untuk menjawab pertanyaan berikut :




**TOP 10 product name apa saja yg paling populer dari masing-masing territory?**

In [ ]:
#Tulis Query disini

%%sql
with product_ranking as(
select
dt.territory_id ,
dp.product_name,
count(fc.product_id) as total_order,
dense_rank() over (partition by dt.territory_id
    order by count(fc.product_id) desc) as ranking
from dim_product dp
join fact_sales fc on fc.product_id = dp.product_id
join dim_territory dt on fc.territory_id = dt.territory_id
group by 1,2
)
select * from product_ranking where ranking <= 10

 * postgresql://tttqkwhz:***@mel.db.elephantsql.com:5432/tttqkwhz
92 rows affected.


[(1, 'HL Mountain Tire', 39, 1),
 (1, 'Patch Kit/8 Patches', 34, 2),
 (1, 'Mountain Tire Tube', 28, 3),
 (1, 'Road-150 Red, 62', 26, 4),
 (1, 'Road-150 Red, 48', 21, 5),
 (1, 'Road-150 Red, 56', 19, 6),
 (1, 'Road-150 Red, 44', 17, 7),
 (1, 'Road-150 Red, 52', 16, 8),
 (1, 'Fender Set - Mountain', 16, 8),
 (1, 'Water Bottle - 30 oz.', 13, 9),
 (1, 'Road Tire Tube', 11, 10),
 (1, 'Mountain-200 Black, 46', 11, 10),
 (1, 'Sport-100 Helmet, Blue', 11, 10),
 (4, 'HL Mountain Tire', 48, 1),
 (4, 'Patch Kit/8 Patches', 42, 2),
 (4, 'Road-150 Red, 48', 37, 3),
 (4, 'Fender Set - Mountain', 37, 3),
 (4, 'Mountain Tire Tube', 36, 4),
 (4, 'Road-150 Red, 56', 31, 5),
 (4, 'Road-150 Red, 62', 30, 6),
 (4, 'Road-150 Red, 52', 29, 7),
 (4, 'Road-150 Red, 44', 23, 8),
 (4, 'Road Tire Tube', 23, 8),
 (4, 'Mountain-200 Black, 46', 20, 9),
 (4, 'Water Bottle - 30 oz.', 19, 10),
 (5, 'Water Bottle - 30 oz.', 1, 1),
 (5, 'Mountain Bottle Cage', 1, 1),
 (5, 'Patch Kit/8 Patches', 1, 1),
 (5, 'Mountain-200 Silver, 38', 1, 1),
 (5, 'Classic Vest, L', 1, 1),
 (6, 'Road Tire Tube', 83, 1),
 (6, 'Patch Kit/8 Patches', 79, 2),
 (6, 'Mountain Tire Tube', 54, 3),
 (6, 'Touring Tire Tube', 49, 4),
 (6, 'Sport-100 Helmet, Black', 34, 5),
 (6, 'ML Road Tire', 34, 5),
 (6, 'Water Bottle - 30 oz.', 32, 6),
 (6, 'HL Road Tire', 32, 6),
 (6, 'Sport-100 Helmet, Blue', 32, 6),
 (6, 'LL Mountain Tire', 30, 7),
 (6, 'AWC Logo Cap', 27, 8),
 (6, 'Sport-100 Helmet, Red', 25, 9),
 (6, 'Bike Wash - Dissolver', 25, 9),
 (6, 'HL Mountain Tire', 24, 10),
 (7, 'Water Bottle - 30 oz.', 58, 1),
 (7, 'Road Bottle Cage', 39, 2),
 (7, 'Sport-100 Helmet, Blue', 36, 3),
 (7, 'Mountain Bottle Cage', 32, 4),
 (7, 'Mountain-200 Black, 42', 29, 5),
 (7, 'Sport-100 Helmet, Red', 28, 6),
 (7, 'Mountain-200 Black, 46', 27, 7),
 (7, 'Mountain-200 Silver, 46', 24, 8),
 (7, 'Patch Kit/8 Patches', 22, 9),
 (7, 'Mountain-200 Silver, 38', 21, 10),
 (8, 'Water Bottle - 30 oz.', 52, 1),
 (8, 'Mountain Bottle Cage', 40, 2),
 (8, 'AWC Logo Cap', 34, 3),
 (8, 'Mountain-200 Silver, 46', 27, 4),
 (8, 'Sport-100 Helmet, Black', 26, 5),
 (8, 'Sport-100 Helmet, Red', 25, 6),
 (8, 'Road Bottle Cage', 24, 7),
 (8, 'Sport-100 Helmet, Blue', 24, 7),
 (8, 'Fender Set - Mountain', 18, 8),
 (8, 'Mountain-200 Silver, 38', 17, 9),
 (8, 'Mountain-200 Black, 46', 17, 9),
 (8, 'Touring Tire', 15, 10),
 (8, 'Mountain-200 Silver, 42', 15, 10),
 (9, 'Water Bottle - 30 oz.', 124, 1),
 (9, 'Mountain Bottle Cage', 76, 2),
 (9, 'Mountain-200 Black, 42', 71, 3),
 (9, 'Mountain-200 Silver, 38', 69, 4),
 (9, 'Sport-100 Helmet, Red', 68, 5),
 (9, 'Fender Set - Mountain', 66, 6),
 (9, 'Sport-100 Helmet, Blue', 63, 7),
 (9, 'Road Bottle Cage', 63, 7),
 (9, 'Mountain-200 Silver, 42', 58, 8),
 (9, 'Patch Kit/8 Patches', 57, 9),
 (9, 'Mountain Tire Tube', 51, 10),
 (10, 'Water Bottle - 30 oz.', 64, 1),
 (10, 'Sport-100 Helmet, Blue', 42, 2),
 (10, 'Sport-100 Helmet, Red', 41, 3),
 (10, 'AWC Logo Cap', 41, 3),
 (10, 'Mountain Bottle Cage', 38, 4),
 (10, 'Road Bottle Cage', 36, 5),
 (10, 'Sport-100 Helmet, Black', 23, 6),
 (10, 'Mountain-200 Black, 42', 22, 7),
 (10, 'Mountain-200 Silver, 42', 21, 8),
 (10, 'Touring Tire', 20, 9),
 (10, 'Mountain-200 Black, 46', 20, 9),
 (10, 'Fender Set - Mountain', 18, 10),
 (10, 'Mountain-200 Black, 38', 18, 10),
 (10, 'Patch Kit/8 Patches', 18, 10)]

**TOP 10 product name apa saja yg paling populer dari masing-masing territory?**\
Jawaban: ada pada tabel

## Question 4


Karena sudah lulus probation nih.. kamu diminta untuk membuat sendiri analysis untuk si Bos. Beberapa step yg perlu dilakukan adalah :
1. Membuat objective analysis (contoh : Karena next campaign mau dibuat maksimal , kita akan menghilight top2 product)
2. Membuat business question (contoh : Apa saja product yg paling populer dari segi jumlah transaksi?)
3. Build the Query untuk menjawab pertanyaan pada poin no 2
4. Tulis Actionable knowledge atau next action setelah mendapat jawaban dari poin 3 (contoh : product abc yg paling tinggi sehingga kita akan fokus untuk promosi product tersebut)




Objective: Perusahaan ingin memahami produk mana yang paling banyak dibeli oleh pelanggan berdasarkan gender, sehingga dapat menyesuaikan strategi pemasaran.

Business Question : Top 10 produk yang paling banyak dibeli oleh pelanggan berdasarkan gender?

In [ ]:
#Tulis Query disini

%%sql
with product_gender_rank as(
SELECT dc.gender, dp.product_name, SUM(fc.quantity) AS jumlah_pembelian,
dense_rank() over (partition by dc.gender
    order by SUM(fc.quantity) desc) as ranking
FROM dim_customer dc
JOIN fact_sales fc ON dc.customer_id = fc.customer_id
JOIN dim_product dp ON fc.product_id = dp.product_id
GROUP BY dc.gender, dp.product_name)

select *
from product_gender_rank
where ranking <= 10

 * postgresql://tttqkwhz:***@mel.db.elephantsql.com:5432/tttqkwhz
23 rows affected.


[('F', 'Water Bottle - 30 oz.', 197, 1),
 ('F', 'Patch Kit/8 Patches', 129, 2),
 ('F', 'Mountain Bottle Cage', 127, 3),
 ('F', 'Sport-100 Helmet, Blue', 116, 4),
 ('F', 'Sport-100 Helmet, Red', 113, 5),
 ('F', 'HL Mountain Tire', 103, 6),
 ('F', 'AWC Logo Cap', 102, 7),
 ('F', 'Road Bottle Cage', 100, 8),
 ('F', 'Mountain Tire Tube', 95, 9),
 ('F', 'Fender Set - Mountain', 95, 9),
 ('F', 'Sport-100 Helmet, Black', 94, 10),
 ('M', 'Water Bottle - 30 oz.', 166, 1),
 ('M', 'Patch Kit/8 Patches', 133, 2),
 ('M', 'Mountain Tire Tube', 105, 3),
 ('M', 'Sport-100 Helmet, Blue', 105, 3),
 ('M', 'Mountain Bottle Cage', 101, 4),
 ('M', 'HL Mountain Tire', 96, 5),
 ('M', 'Fender Set - Mountain', 93, 6),
 ('M', 'Road Bottle Cage', 91, 7),
 ('M', 'Sport-100 Helmet, Red', 88, 8),
 ('M', 'AWC Logo Cap', 88, 8),
 ('M', 'Sport-100 Helmet, Black', 81, 9),
 ('M', 'Road Tire Tube', 80, 10)]

Actionable knowledge :
1. Dari tabel diatas, produk-produk seperti "Water Bottle - 30 oz.," "Patch Kit/8 Patches," "Mountain Bottle Cage," dan "Sport-100 Helmet, Blue" terlihat populer di antara kedua jenis kelamin. Dari keadaan tersebut bisa dilakukan promosi khusus seperti couple produk salah satu atau lebih dari keempat produk tersebut.
2. Produk "Sport-100 Helmet, Red" lebih populer terjual kepada jenis kelamin perempuan, sehingga bisa dilakukan promosi khusus terhadap produk ini kepada perempuan.